### PR TITLE
Fix the complete book!

### LINKED ISSUE #1647: Update book to use new API after rendy rewrite
The current book on master needs to be updated to actually work with the code on master.

This issue will track progress & assignees for all the chapter as this will require some effort.

## Steps

How to work on a book chapter:

1. Temporarily remove the other chapters from `book/src/SUMMARY.md`
2. `cargo clean && cargo build --all --features "gltf vulkan"`
3. Use `mdbook test book -L target/debug/deps` to check for compilation issues
4. Read the main file (`book/src/chapter_name.md`) and the sub-chapters (`book/src/chapter_name/*.md`) and check them for accuracy
5. Open a PR with your fixes, discarding changes in `book/src/SUMMARY.md`

## Progress

- 🙏 - to do
- 🚧 - chapter is being worked on
- ✔️ - chapter is correct

| 🏴󠁧󠁢󠁮󠁩󠁲󠁿 | Chapter | Assignee |
| - | - | - |
| ✔️ | animation | n/a |
| ✔️ | appendices | @azriel91 |
| ✔️ | assets | @torkleyy |
| ✔️ | concepts | @torkleyy |
| ✔️ | controlling_system_execution | @hbina |
| ✔️ | getting_started | @torkleyy |
| ✔️ | pong-tutorial | (see sub chapters below) |
| ✔️ | pong-tutorial - Creating a window | @belovachap |
| ✔️ | pong-tutorial - Drawing the paddles | @Frizi |
| ✔️ | pong-tutorial - Making a ball move and bounce | @Frizi |
| ✔️ | pong-tutorial - Winning rounds | @Frizi |
| ✔️ | prefabs | @torkleyy |
| 🚧 | sprites | @torkleyy |
| 🚧 | testing | @torkleyy |

### PR DESCRIPTION
**Pull request by torkleyy**
## Description

* Made the complete book pass `mdbook test`

Fixes #1647 

## PR Checklist

By placing an x in the boxes I certify that I have:

- [ ] Ran `cargo test --all` locally if this modified any rs files.
- [ ] Ran `cargo +stable fmt --all` locally if this modified any rs files.
- [ ] Updated the content of the book if this PR would make the book outdated.
- [ ] Added a changelog entry if this will impact users, or modified more than 5 lines of Rust that wasn't a doc comment.
- [ ] Added unit tests for new APIs if any were added in this PR.
- [x] Acknowledged that by making this pull request I release this code under an MIT/Apache 2.0 dual licensing scheme.

### CODE REVIEW

File: book/src/sprites/display_the_texture.md [hunk 1]
```diff
@@ -3,31 +3,31 @@
 With a [`DrawFlat2D`][doc_drawflat2d] render pass set up and a loaded texture, it's already possible to render the full texture. The [`TextureHandle`][doc_tex_handle] itself implements [`Component`][doc_component], so you can just attach the [`Handle`][doc_handle] to an entity and it'll show up!
 
 For anything rendered by the [`DrawFlat2D`][doc_drawflat2d] pass, it's also possible to optionally attach a [`Flipped`][doc_flipped] component to the entity, which will signal to the renderer that you want to flip your texture horizontally or vertically when rendering.
+TODO
```
Comment (torkleyy) [inline book/src/sprites/display_the_texture.md]: I didn't how this works now.

File: book/src/sprites/load_the_texture.md [hunk 2]
```diff
@@ -38,11 +37,14 @@ impl SimpleState for ExampleState {
...
+cloneable reference to the texture.
 
     When you use [`loader.load(..)`][doc_load] to load an [`Asset`][doc_asset], the method returns immediately with a unique handle for your texture. The actual asset loading is handled asynchronously, so if you attempt to use the texture handle to retrieve the texture, such as with [`world.read_resource::<AssetStorage<Texture>>()`][doc_read_resource][`.get(texture_handle)`][doc_asset_get], you will get a `None` until the `Texture` has finished loading.
 
-The loaded texture will use linear filter, e.g. screen pixels will be linearly interpolated between the closest image pixels. In layman's terms, if your images have small resolution, sprites will look blury. Use `TextureMetadata::srgb_scale()` instead to avoid such effect. Screen pixel will be taken from nearest pixel of texture in that case.
+The loaded texture will use linear filter, e.g. screen pixels will be linearly interpolated between the closest image pixels. In layman's terms, if your images have small resolution, sprites will look blury.
+Use `TextureMetadata::srgb_scale()` instead to avoid such effect. Screen pixel will be taken from
+ nearest pixel of texture in that case. TODO
```
Comment (torkleyy) [inline book/src/sprites/load_the_texture.md]: This needs to be changed.

File: book/src/sprites/display_the_texture.md (hunk 1 above)
Comment (azriel91) [reply book/src/sprites/display_the_texture.md]: Flipping sprites is now done in one of two methods:

* By passing `true` to `Sprite::from_pixel_values` for the `flip_horizontal` or `flip_vertical` parameters when building the sprite sheet.
* At runtime, by changing the y or x rotation on the entity's `Transform`.

    ```rust
    use amethyst::core::math::RealField;

    transform.set_rotation_y_axis(f32::pi()); // Flip horizontally
    transform.set_rotation_x_axis(f32::pi()); // Flip vertically
    ```

Comment (azriel91) [reply book/src/sprites/display_the_texture.md]: It doesn't really suit the current flow of the text to mention sprite sheet loading though.

File: book/src/sprites/orthographic_camera.md [hunk 3]
```diff
@@ -36,17 +36,20 @@ impl ExampleState {
         let mut transform = Transform::default();
         transform.set_translation_xyz(0., 0., 10.);
```
Comment (azriel91) [inline book/src/sprites/orthographic_camera.md]: the `y` value should also be `height` (otherwise it renders a blank screen)

File: book/src/sprites/set_up_the_render_pass.md [hunk 4]
```diff
@@ -1,55 +1,96 @@
...
+// We pass the subpass builder a description of our groups for construction
+let pass = graph_builder.add_node(
+    SubpassBuilder::new()
+        .with_group(DrawFlat2DDesc::default().builder()) // Draws sprites
+        .with_color(color)
+        .with_depth_stencil(depth)
+        .into_pass(),
+);
+ 
+         // Finally, add the pass to the graph
```
Comment (azriel91) [inline book/src/sprites/set_up_the_render_pass.md]: probably want to hide these lines as well?

File: book/src/assets/how_to_define_custom_formats.md [hunk 5]
```diff
@@ -2,9 +2,15 @@
 
 This guide explains how to define a new asset format. This will allow Amethyst to load assets stored in a particular encoding.
 
-There are two formatting traits in Amethyst &ndash; `Format<A: Asset>` and `SimpleFormat<A: Asset>`. `Format` requests a loading implementation that provides detection when an asset should be reloaded for [hot reloading][doc_hrs]; `SimpleFormat` requires a loading implementation. A blanket implementation will implement `Format` for all `SimpleFormat` implementations, and provides hot reloading detection based on file modification time &ndash; this is sufficient for most applications that require hot reloading. This guide covers implementation of the `SimpleFormat` trait.
+There is a trait in Amethyst for implementing a format: `Format<A: Asset::Data>`.
```
[the remark this answers is not in the document]
Comment (Frizi) [reply book/src/assets/how_to_define_custom_formats.md]: Well, the "true" syntax would really be `impl<D, A> Format<D> where A: Asset, D: A::Data`, but that is underconstrained. It means the asset type must be specified. Maybe explaining it like `Format<Data>`, with additional specific example, like `Format<<Texture as Asset>::Data>`. If this is too complicated, i would go with simple `Format<Asset::Data>`.

File: book/src/assets/how_to_define_custom_formats.md [hunk 6]
```diff
@@ -40,20 +48,17 @@ If you are defining a new format that may be useful to others, [please send us a
     #[derive(Clone, Copy, Debug, Default)]
     pub struct MyLangFormat;
 
-    impl<A> SimpleFormat<A> for MyLangFormat
+    impl<D> Format<D> for MyLangFormat
```
Comment (Frizi) [inline book/src/assets/how_to_define_custom_formats.md]: In the light of changes that happened to formats, i'm not sure if this is really the thing we should be explaning first. Generic formats are kinda tricky, as those are basically a proxy for serialization format like ron or json. It's kinda a hack really. The second, i would call "more idiomatic" group of formats are actually solving the problem "i have some data and i want to transform it to soemthing that can be loaded". Those formats specify their "output type" directly, like ObjFormat that is converting a blob with obj data into MeshBuilder. This is what i would consider you want to add more often, like FbxFormat or BmpFormat could be a thing.

But then, do we even care at this point? This will likely change again once atelier is in place, so maybe we defer that.

Comment (torkleyy) [reply book/src/assets/how_to_define_custom_formats.md]: For now, I just want to fix the book to be correct. I think I agree this can be improved, but I'll just leave it like this for now since the scope of this PR is only to get the book correct.

**Comment by Frizi** [comment]
bors r=Frizi,azriel91

**Comment by torkleyy** [comment]
bors ping

**Comment by torkleyy** [comment]
bors r=Frizi,azriel91

**Comment by torkleyy** [comment]
bors r=Frizi,azriel91

**Comment by Frizi** [comment]
bors retry

**Comment by Frizi** [comment]
bors retry
